1394 add eye tracking to behavior ophys session

njmei opened the pull request:
This PR adds the ability for internal users to access DeepLabCut eye tracking ellipse fit data from the behavior project cache.

Notes:
The following columns are not currently added in the output dataframe obtained from the project cache.
- blind/outlier corrected area (NaN where likely_blink == True): This column was not added because it is too similar to existing area columns and in the spirit of the AllenSDK should probably be something that an end user should have control over.
- normalized corrected area: In a similar vein, because there are a *lot* of ways to perform this, I think it's best to leave this column off of the 'basic' eye tracking dataframe.

That said, I've added an example notebook in cd56612 that demonstrates how to generate the above two columns.

Relates to: #1394

dougollerenshaw commented:
@njmei mind adding @matchings as a reviewer?

dougollerenshaw commented on allensdk/brain_observatory/behavior/eye_tracking_processing.py:
```diff
@@ -0,0 +1,164 @@
+
+def compute_pupil_area(df_row: pd.Series) -> float:
+    # Calculate the area of the pupil as a circle using the max of the
+    # height/width as radius
```
I might suggest fleshing out this comment a bit. Add:
# This assumes that the pupil is a perfect circle and any eccentricity is a result of the angle at which the pupil is being viewed

dougollerenshaw commented on allensdk/brain_observatory/behavior/eye_tracking_processing.py:
```diff
@@ -0,0 +1,164 @@
+    return np.pi * max_dim * max_dim
+
+
+def compute_eye_area(df_row: pd.Series) -> float:
```
Add a comment/docstring here. Note that this applies to both the eye fit and the corneal reflection fit.

dougollerenshaw commented:
Another general comment/question: How can we make it obvious to users what the units/coordinate space of these values are? It should be obvious that:

* positions are in units of pixels.
* areas are in units of pixels^2
* all values are in the coordinate space of the eye tracking camera, NOT the coordinate space of the stimulus display (i.e., this is not gaze location), with 0,0 being the upper-left corner of the image.

dougollerenshaw commented on allensdk/brain_observatory/behavior/eye_tracking_processing.py:
```diff
@@ -0,0 +1,164 @@
+    return eye_tracking_data.astype(float)
+
+
+def determine_outliers(eye_tracking_df: pd.DataFrame,
```
I don't completely understand how this will work. Will it check *every* column for outliers and return TRUE if *any* value exceeds a z-score of `z_threshold`? 

**I think it should only operate on the area columns**. In the equivalent VBA implementation, I discovered that transient eye movements can be sufficiently large and rare such that they have a large z-score, but they are still perfectly captured by the ellipse fits. Also, since the `outliers` column is used in the `determine_likely_blinks` method, these large ballistic eye movements end up being labeled as likely_blink == True, which is not correct.

njmei replied:
This currently operates on *all* columns. As my thinking was that extremely large z-scores in any of the parameters would suggest something had gone wrong with the tracking. Given that you've looked and seen that ellipse fits are still accurate, I can change this to only operate on `eye_area` and `pupil_area` columns.

dougollerenshaw commented on allensdk/brain_observatory/behavior/behavior_ophys_session.py:
```diff
@@ -305,6 +306,21 @@ def segmentation_mask_image(self) -> Image:
+        parameters for eye, pupil, and corneal reflection.
+
+        :rtype: pandas.DataFrame
+        """
```
I think this docstring should be expanded quite a bit to explain the following:

* all columns starting with 'pupil_' represent ellipse fit parameters relating to the pupil
* all columns starting with 'eye_' represent ellipse fit parameters relating to the eyelid
* all columns staring with 'cr_' represent ellipse fit parameters relating to the corneal reflection, which is caused by an infrared LED positioned near the eye tracking camera.
* positions are in units of pixels.
* areas are in units of pixels^2
* all values are in the coordinate space of the eye tracking camera, NOT the coordinate space of the stimulus display (i.e., this is not gaze location), with 0,0 being the upper-left corner of the image.
* the 'likely_blink' column is True for any row where pupil fit failed OR eye fit failed OR an outlier fit was identified
* all ellipse fits are derived from tracking points that were output by DeepLabCut model that was trained on hand-annotated data from a subset of imaging sessions on the optical physiology rigs. The raw tracking points are not available [but should the be?]

dougollerenshaw commented on allensdk/brain_observatory/behavior/behavior_ophys_session.py:
```diff
@@ -305,6 +306,21 @@ def segmentation_mask_image(self) -> Image:
+    @property
+    def eye_tracking_data(self) -> pd.DataFrame:
+        """A dataframe containing DeepLabCut eye tracking ellipse fit
+        parameters for eye, pupil, and corneal reflection.
```
I'd suggest rewording this part of the docstring slightly to be a little more precise about what this represents (it's not actually the output from DeepLabCut - it contains ellipse fit parameters *derived from* the DeepLabCut model output). How about "A dataframe containing ellipse fit parameters for the eye, pupil and corneal reflection. Fits are derived from tracking points from a DeepLabCut model applied to video frames of the mouse's right eye. Raw tracking points and raw video frames are not exposed by the SDK"

njmei commented:
@dougollerenshaw Many thanks for the detailed feedback! I will try to get all this implemented by the end of today!

dougollerenshaw commented:
No problem. Nice work on this!

njmei commented:
- [ ] squash commits once things are okayed

njmei commented:
@dougollerenshaw Hi Doug, I've addressed the points you've brought up and also discovered a bug in my outlier detection implementation (that I've since fixed). Let me know if any other changes are desired!

kschelonka commented on allensdk/brain_observatory/behavior/behavior_ophys_session.py:
```diff
@@ -305,6 +306,45 @@ def segmentation_mask_image(self) -> Image:
+        return self._eye_tracking_data
+
+    @eye_tracking_data.setter
+    def eye_tracking_data(self, value):
```
You have some internal methods that have configuration variables (e.g. `determine_outliers` has `z_threshold`, and `determine_likely_blinks` has `dilation_frames`). I can see a use-case where the user would want to have control over these variables, but there's no way to actually change them from the defaults either by configuring the API or through the session. You should include a way to set them from the session somewhere.

njmei replied:
Good idea, do you think changing `eye_tracking_data` from a property to a method like `get_eye_tracking_data` for the behavior+ophys session object would be a reasonable way to expose those parameters?

kschelonka replied:
I'd like to keep the access pattern the same as what's already been done so that it's not confusing for users. I think maybe instead we could add these arguments as optional parameters when the behavior session is initialized -- either as keyword arguments or as a dictionary of keyword arguments. They would be stored as session attributes. We'd update the logging to provide users with logs at the `info` level to clarify which arguments were used when `eye_tracking_data` is accessed via the session object.

kschelonka replied:
The only part that might muddy the waters about this is that if users update the attributes, they won't re-pull the eye tracking data since it'll already be cached. We'll need to decide how to deal with this (require users to create a new session object since technically everything should be read-only? use get/set logic on the session config vars to know when to "refresh" the data? etc.)

dougollerenshaw replied:
I don't have any opinions/ideas on the implementation details, but I do agree with @kschelonka that it would be ideal for users to be able to control these parameters.

kschelonka replied:
I think since we don't necessarily want to re-pull all the data just from changing these configs, I suggest that we set these as properties, and use the setter method to set the value for `self._eye_tracking_data` to `None`. That will trigger pulling that data (and that data only) again. Include enough logging to make it clear to users what is happening.

njmei replied:
@kschelonka @dougollerenshaw Okay, I've updated this PR to add the described feature. Let me know if any other changes are desired!

dougollerenshaw replied:
@njmei: Either the parameter change isn't working, or I'm not using it properly. Following I used your notebook, then added a cell to plot the likely_blink column vs. time for two different settings of dilation_frames. But the result didn't change. Here's the code:

```
fig,ax=plt.subplots()
session.eye_tracking_processing_args = {"z_threshold": 1.0, "dilation_frames": 2}
eye_data1 = session.eye_tracking_data.copy(deep=True)
ax.plot(
    eye_data1['time'],
    eye_data1['likely_blink'],
    color = 'black'
)

session.eye_tracking_processing_args = {"z_threshold": 1.0, "dilation_frames": 4}
eye_data2 = session.eye_tracking_data.copy(deep=True)
ax.plot(
    eye_data2['time'],
    eye_data2['likely_blink'],
    color = 'red',
    linewidth=4,
    alpha=0.5,
    linestyle=':'
)

ax.legend(['dilation frames = 2', 'dilation frames = 4'])
ax.set_xlim(2576,2581)
```

And here's the result (I'd expect the red-dashed line to rise sooner and fall later relative to the black line, but instead they are identical)
![image](https://user-images.githubusercontent.com/19944442/76805452-0b474280-679c-11ea-9a85-1fdc86ea6f2e.png)

I'm getting the notification that the underlying parameter is changing, so it looks like something is happening in the background:
![image](https://user-images.githubusercontent.com/19944442/76805559-595c4600-679c-11ea-8995-93aa2ca40680.png)

njmei replied:
@dougollerenshaw Doh! I forgot to commit a change, so wasted time not being able to reproduce your example. I've updated the examples notebook to demonstrate the correct behavior.

dougollerenshaw replied:
Thanks @njmei! I'm seeing a couple weird things now.
1) When I run the exact same code block as above, the black line (for dilation_frames=2) is different than before. Why would that one have changed?

Here's the plot:
![image](https://user-images.githubusercontent.com/19944442/76810979-3b4b1180-67ad-11ea-8cf7-f3177ef09b54.png)

dougollerenshaw replied:
2) If I set dilation_frames = 0, then *every* frame gets labeled as a likely blink:
![image](https://user-images.githubusercontent.com/19944442/76811062-75b4ae80-67ad-11ea-86e6-6823f38cf25e.png)
![image](https://user-images.githubusercontent.com/19944442/76811101-92e97d00-67ad-11ea-8ac8-545fd3174a26.png)

dougollerenshaw replied:
3) If I set dilation_frames=1, it's not clear to me what's triggering the likely_blink column to be set to True. For example, here's an example where indices 20231 to 20233 are labeled likely_blink=True. I'd expect the middle value (index 20232) to have a NaN or outlier fit. It doesn't:
![image](https://user-images.githubusercontent.com/19944442/76811279-376bbf00-67ae-11ea-9734-2edb4feda95e.png)
What did the algorithm identify that caused it to label these three frames as likely_blink=True?

njmei replied:
@dougollerenshaw 

1. Because of the commit I forgot, setting `eye_tracking_processing_args` was not actually doing anything and you were likely seeing the results of eye tracking data processed with defaults: `{"z_threshold": 3.0, "dilation_frames": 2}`

2. Because dilation is accomplished with the `scipy.ndimage.binary_dilation` function the documentation states that: [iterations int, optional

    The dilation is repeated iterations times (one, by default). If iterations is less than 1, the dilation is repeated until the result does not change anymore. Only an integer of iterations is accepted.
](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_dilation.html)

I can change this behavior so that a `dilation_frame=0` does no `binary_dilation` if desired.

3. If any of `cr_area`, `eye_area`, or `pupil_area` are outliers then that would cause them to be labeled as a `likely_blink`.

dougollerenshaw replied:
Thanks @njmei!

1) You're right, when I changed back to `z_threshold=3`, everything acts as expected (I added dot markers to make it clear that the red line rises two frames earlier and falls two frames later than the black line, as expected):
![image](https://user-images.githubusercontent.com/19944442/76812145-35572f80-67b1-11ea-87ba-5c91fb5bb504.png)

2) Thanks for digging into the documentation. That's very weird behavior for `n_iterations=0`! I agree with your proposal: If `dilation_frames=0`, do not call `binary_dilation`.

3) Yes, when I set `z_threshold=3`, that made `likely_blink=False` for index 20232 (and its neighbors). So you were likely right that it was the corneal reflection fit that triggered that, given the low threshold (`z_threshold=1`) for outliers.

kschelonka commented on allensdk/brain_observatory/behavior/behavior_ophys_session.py:
```diff
@@ -56,9 +56,14 @@ def __init__(self, api=None):
+
+        # Keep track of args for data processing functions that hydrate
+        # ophys behavior session properties.
+        self._eye_tracking_processing_args = None
```
Please also update the class constructors (I think only LIMS in this case, since this isn't relevant to NWB that's already processed?) and `__init__` method for the session. I think it's confusing if these parameters can only be set after the class instance is created.

Also is there a reason why you packaged these up together rather than just using the names of the arguments in the downstream functions?

njmei replied:
1. I don't believe NWB files exist for behavior + ophys yet, but can definitely update the LIMS class constructor and the __init__ method.

2. Are you asking why I packaged the args as a dict instead of having individual properties with setters? If so, I wanted to avoid too much property namespace pollution. If some future property has processing that can accept a bunch of params, that would be a lot of boilerplate to add.

wbwakeman replied:
There definitely are NWB for ophys + behavior. That is what the BEHAVIOR_OPHYS_WRITE_NWB_QUEUE does.
allensdk.brain_observatory.behavior.write_nwb
(These have not been publicly released yet though)

kschelonka replied:
Hi @wbwakeman , do you know if the eye tracking is stored in NWB in a raw enough format to make use of these methods, or is it already packaged? If you don't know offhand that's something we should look into @njmei

wbwakeman replied:
Sorry; I don't know offhand

njmei replied:
@kschelonka I took a quick look yesterday and it looks like the eye tracking ellipse fits are not currently stored in the NWB file. That said, it looks like it would be a separate scope of work:

https://github.com/AllenInstitute/AllenSDK/issues/1391

kschelonka replied:
I see your point about not wanting clutter, but to me it's pretty confusing if I have to pass the whole dictionary for all the eye arguments just to change one value. One other option could be going the more OO way of scikit-learn, spark transformers, etc. which have get and set param methods. Then we could handle everything in `set_params` and not worry about using properties to control it.

kschelonka replied:
I can speak firsthand from using the caiman package that the dictionary method is very confusing for the user. In the case of caiman for example I knew the name of the arguments I wanted, but I didn't know exactly where they were stored in their nested param dictionary. Right now it seems simple (just eye tracking), but it can quickly get more complex. With this proposed method we are requiring users to not only know the keys of the arguments, but also the names of the dictionary groupings to which they belong.

njmei replied:
Okay, this is done. Feel free to unresolve if anything is still unsatisfactory.

kschelonka commented on allensdk/brain_observatory/behavior/behavior_ophys_session.py:
```diff
@@ -305,6 +310,88 @@ def segmentation_mask_image(self) -> Image:
+            "'eye_tracking_processing_args' have been changed! "
+            "The session 'eye_tracking_data' will now be updated."
+        )
+        _ = self.eye_tracking_data
```
`self._eye_tracking_data = None`

njmei replied:
Yeah, I forgot to commit this change. It's in now.

kschelonka replied:
Since that's what the eye tracking property is checking to see if it needs to pull data from the api:
```
if self._eye_tracking_data is None:
        self._eye_tracking_data = self.api.get_eye_tracking_data(**kwargs)
        return self._eye_tracking_data
```

njmei commented:
- [ ] add tests for params mixin and ophys_session
- [ ] rebase and squash this chonker of a PR

njmei commented:
@kschelonka @dougollerenshaw Okay, hopefully third time is the charm ;)

For some odd reason, circleci has decided to start running tests for this PR even though don't have CircleCI set up for the AllenSDK so ignore it's failure for now...

kschelonka commented on allensdk/test/brain_observatory/test_session_api_utils.py:
```diff
@@ -0,0 +1,227 @@
+     False)
+
+], indirect=["mixin_harness_fixture"])
+def test_needs_data_refresh(mixin_harness_fixture, params_to_set, data_params, expected):
```
Please add some tests for setting more than one parameter at once, and different types (bool, int, list, array, etc.)

njmei replied:
@kschelonka Many thanks for the suggestion, it definitely encountered some problems that I've hopefully addressed in the commits below!

njmei commented:
- [ ] Wait for rc/1.7.0
- [ ] Squash commits
- [ ] Add What's new blurb

dougollerenshaw commented on allensdk/brain_observatory/behavior/eye_tracking_processing.py:
```diff
@@ -0,0 +1,219 @@
+        eye tracking frames.
+    """
+
+    if len(frame_times) != len(eye_data.index):
```
@njmei there are sessions where apparently spurious  edges in the sync file causes there to be a mismatch between the sync count and the video frame count. It's good that you will raise an error in this case. Colin F (I can't find his github handle!) and I were just discussing this separately and he might have some ideas on dealing with this issue.
